torchquad/integration/newton_cotes.py

```diff
@@ -47,8 +48,16 @@ def calculate_result(self, function_values, dim, n_per_dim, hs):
             backend tensor: Quadrature result
         """
         # Reshape the output to be [N,N,...] points instead of [dim*N] points
-        function_values = function_values.reshape([n_per_dim] * dim)
-
+        self.integrand_shape = function_values.shape[1:]
+        if len(self.integrand_shape) == 0 or (len(self.integrand_shape) == 1 and self.integrand_shape[0] == 1):
+            function_values = function_values.reshape([n_per_dim] * dim)
+        else:
+            dim_shape = [n_per_dim] * dim
+            new_shape = [*self.integrand_shape, *dim_shape]
+            einsum = "".join([chr(i + 65) for i in range(len(function_values.shape))])
+            function_values = anp.einsum(f'{einsum}->{einsum[1:]}{einsum[0]}', function_values)
+            function_values = function_values.reshape(new_shape)
+            assert new_shape == list(function_values.shape)
```

Comment:
Necessary?

Reply:
I don't quite get what you are doing here. To explain the original code:

1. For speed, we evaluate all points on which we evaluate the integrand at once (this utilizes the pytorch (or whichever framework) internal vectorization). For that reason the tensor passed to this function has shape [N^dim]. 
2. Now, to compute the integral we need to reshape it into [N,N,N,...]

I suspect, all that should be necessary here, is to change

`function_values = function_values.reshape([n_per_dim] * dim)`

into something like

`function_values = function_values.reshape([n_integrals]+[n_per_dim] * dim)`

So, if you evaluate, e.g. a 2D integrand 10 times on a 5x5 grid, you would receive a shape [250] input tensor, reshape into [10,5,5] and then in composite rule apply the rule simultaneously on the 10 5x5 grids.

Makes sense?

Reply:
I think you're right - the issue is the einstein sum, it seems?  Since it requires these extra steps that make the reshape call different?  I think I should be able to simplify this.

Reply:
Yes, I think, this isn't quite necessary. All I would do here is reshape into from (M*N^dim) into M,N,N,N and then modify the composite rule to ignore the leading dimension if there is one?

Reply:
Ok @gomezzz maybe we do need a zoom call.  I think I understand what you're saying. But the issue is that, I think, that is what I'm doing.  I have the new shape as `new_shape = [*self.integrand_shape, *dim_shape]` which seems similar to what you are saying.  But the issue is that the `evaluate_integrand` expects the `N` of the integration dimension to be in the first entry.  We could change this , but this also seemed like not a good idea.

Reply:
See https://github.com/esa/torchquad/blob/099f7f566da0661f16362cf6cda478a6b34f5a4a/torchquad/integration/base_integrator.py#L54-L70

Reply:
i.e we need the `einsum` to put it in the last entry instead, so then it can be easily reshaped.  I could be wrong, but just reshaping has unintended consequences.

Reply:
For example, just `reshape` as you suggest at the moment causes things to fail.

Reply:
It's also possible that I am just not getting your code :D But let's have a call, surely quicker. Just write me an email to pablo.gomez at esa.int with some time slots that work and I'll send an invite 🙏 :)

Reply:
emailed!